QgisModelBaker/libqgsprojectgen/dbconnector/mssql_connector.py

```diff
@@ -0,0 +1,409 @@
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+"""
+
+import re
+import pyodbc
+from pyodbc import (ProgrammingError, InterfaceError)
+from .db_connector import (DBConnector, DBConnectorError)
+
+METADATA_TABLE = 't_ili2db_table_prop'
+METAATTRS_TABLE = 't_ili2db_meta_attrs'
+
+
+class MssqlConnector(DBConnector):
+    def __init__(self, uri, schema):
+        DBConnector.__init__(self, uri, schema)
+
+        try:
+            self.conn = pyodbc.connect(uri)
+        except (ProgrammingError, InterfaceError, pyodbc.Error, pyodbc.OperationalError) as e:
+            raise DBConnectorError(str(e), e)
+
+        self.schema = schema
+        
+        self._bMetadataTable = self._metadata_exists()
+        self.iliCodeName = 'iliCode'
+
+    def map_data_types(self, data_type):
+        result = data_type.lower()
+        if 'timestamp' in data_type:
+            result = self.QGIS_DATE_TIME_TYPE
+        elif 'date' in data_type:
+            result = self.QGIS_DATE_TYPE
+        elif 'time' in data_type:
+            result = self.QGIS_TIME_TYPE
+
+        return result
+    
+    def db_or_schema_exists(self):
+        if self.schema:
+            cur = self.conn.cursor()
+            cur.execute("""
+                SELECT case when count(schema_name)>0 then 1 else 0 end
+                FROM information_schema.schemata
+                where schema_name = '{}'
+            """.format(self.schema))
+
+            return bool(cur.fetchone()[0])
+
+        return False
+
+    def metadata_exists(self):
+        return self._bMetadataTable
+    
+    def _metadata_exists(self):
+        return self._table_exists(METADATA_TABLE)
+
+    def _table_exists(self, tablename):
+        if self.schema:
+            cur = self.conn.cursor()
+            cur.execute("""
+            SELECT count(TABLE_NAME) as 'count'
+                FROM INFORMATION_SCHEMA.TABLES
+                WHERE TABLE_TYPE = 'BASE TABLE'
+                AND TABLE_SCHEMA = '{}'
+                    AND TABLE_NAME = '{}'
+            """.format(self.schema, tablename))
+
+            return bool(cur.fetchone()[0])
+
+        return False
+
+    def get_tables_info(self):
+        res = []
+
+        if self.schema:
+            metadata_exists = self.metadata_exists()
+
+            ln = "\n"
```

Comment:
I tend to prefer to write SQL in Python as follows
```python
           stmt = """
SELECT distinct
    tbls.TABLE_SCHEMA AS schemaname
  , tbls.TABLE_NAME AS tablename
                  """.format(
                      var1, var2, …
                  )

```

I find it easier to read as the SQL pops out but that's up to you.